[ao] Added fx model report per_channel detector

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79243
* __->__ #79104

Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: python test/quantization/fx/test_model_report_fx.py TestModelReportFxDetector

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D37055549](https://our.internmc.facebook.com/intern/diff/D37055549)

vspenubarthi committed:
[ao][work-in-progress] Added fx model report per_channel detector

Summary: This code is still a work in progress. The overall main
    _detect_per_channel code is ready to be checked, however more unit tests
    are still being written. It would be great if the code can be checked at
    a high level for both any glaring issues and any major code formatting
    discrepencies.

Test Plan: test/quantization/fx/test_model_report_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added fx model report per_channel detector"


Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: test/quantization/fx/test_model_report_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added fx model report per_channel detector"


Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: test/quantization/fx/test_model_report_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added fx model report per_channel detector"


Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: test/quantization/fx/test_model_report_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -0,0 +1,142 @@
+            }
+
+        # recurse on children
+        for child in module.children():
```
Instead of recursing on children, can we achieve the same result by using `module.named_modules` without recursion?

HDCharles replied:
I think they both get only the direct children, though named_modules gets name, child rather than just child.

vkuzo replied:
`named_modules` is a generator over all modules in the network, not just children of the current module. I'd recommend removing the recursion unless there is no way to do it without recursion.

HDCharles replied:
oh ya that makes sense

vspenubarthi committed:
Update on "[ao] Added fx model report per_channel detector"


Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: test/quantization/fx/test_model_report_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/fx/_model_report/_detector.py:
```diff
@@ -0,0 +1,143 @@
+}
+
+
+def _detect_per_channel(model: GraphModule) -> Tuple[str, Dict[str, Any]]:
```
nit: this can be a normal module, should still work since we aren't using any fx specific methods

vspenubarthi committed:
Update on "[ao] Added fx model report per_channel detector"


Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: test/quantization/fx/test_model_report_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added fx model report per_channel detector"


Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: test/quantization/fx/test_model_report_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -0,0 +1,433 @@
+
+    """Prepares and callibrate the model"""
+
+    def prepare_model_and_run_input(self, model, q_config_mapping, input):
```
this is a helper function rather than a test so prepend `_`

vspenubarthi committed:
Update on "[ao] Added fx model report per_channel detector"


Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: test/quantization/fx/test_model_report_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added fx model report per_channel detector"


Summary: This code is meant to be a tool to help people get the most out
of their backend by hinting them to use per_channel quantization if it's
supported, which will help increase accuracy significantly. The code is
completed and ready to be reviewed.

Test Plan: test/quantization/fx/test_model_report_fx.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@vspenubarthi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37055549).

vspenubarthi commented:
@pytorchbot merge -g
